File: backend/src/main/java/com/back/back9/global/security/OAuth2SuccessHandler.java
package com.back.back9.global.security;

import com.back.back9.domain.user.entity.User;
import com.back.back9.domain.user.repository.UserRepository;
import com.back.back9.domain.user.service.UserService;
import com.back.back9.domain.wallet.service.WalletService;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Lazy;
import org.springframework.security.core.Authentication;
import org.springframework.security.web.authentication.AuthenticationSuccessHandler;
import org.springframework.stereotype.Component;

import java.io.IOException;

@Component
public class OAuth2SuccessHandler implements AuthenticationSuccessHandler {

    @Lazy
    @Autowired
    private UserService userService;

    @Lazy
    @Autowired
    private WalletService walletService;

    @Lazy
    @Autowired
    private UserRepository userRepository;

    @Value("${spring.profiles.active:dev}")
    private String activeProfile;

    @Value("${app.oauth2.frontend-url}")
    private String frontendUrl;

    @Value("${app.oauth2.cookie-domain:}")
    private String cookieDomain;

    @Override
    public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response, Authentication authentication) throws IOException {
        System.out.println("=== OAuth2SuccessHandler 시작 ===");
        System.out.println("활성 프로파일: " + activeProfile);

        SecurityUser securityUser = (SecurityUser) authentication.getPrincipal();
        User user = securityUser.getUser();

        System.out.println("인증된 사용자: " + user.getUserLoginId());

        // User ID가 null인지 확인
        if (user.getId() == null) {
            System.out.println("사용자 ID가 null이므로 저장 필요");
            user = userRepository.save(user);
            System.out.println("사용자 저장 완료, ID: " + user.getId());
        }

        String accessToken = userService.genAccessToken(user);
        String role = user.getRole().name();

        System.out.println("액세스 토큰 생성 완료");
        System.out.println("사용자 역할: " + role);

        // 지갑이 존재하지 않으면 생성
        if (!walletService.existsByUserId(user.getId())) {
            System.out.println("지갑이 존재하지 않아 생성");
            walletService.createWallet(user.getId());
        } else {
            System.out.println("지갑이 이미 존재함");
        }

        String apiKey = user.getApiKey();

        System.out.println("=== 쿠키 설정 시작 ===");
        System.out.println("accessToken: " + accessToken);
        System.out.println("apiKey: " + apiKey);
        System.out.println("role: " + role);

        // 환경별 쿠키 설정
        boolean isLocal = "dev".equals(activeProfile);

        if (isLocal) {
            // 로컬 환경: 일반 쿠키 설정
            System.out.println("로컬 환경 - 일반 쿠키 설정");

            String accessTokenCookie = String.format(
                    "accessToken=%s; Path=/; HttpOnly=true; SameSite=Lax; Max-Age=3600",
                    accessToken
            );

            String apiKeyCookie = String.format(
                    "apiKey=%s; Path=/; HttpOnly=true; SameSite=Lax; Max-Age=3600",
                    apiKey
            );

            String roleCookie = String.format(
                    "role=%s; Path=/; HttpOnly=true; SameSite=Lax; Max-Age=3600",
                    role
            );

            response.addHeader("Set-Cookie", accessTokenCookie);
            response.addHeader("Set-Cookie", apiKeyCookie);
            response.addHeader("Set-Cookie", roleCookie);
        } else {
            // 프로덕션 환경: 크로스 도메인 쿠키 설정
            System.out.println("프로덕션 환경 - 크로스 도메인 쿠키 설정");

            String accessTokenCookie = String.format(
                    "accessToken=%s; Path=/; HttpOnly=true; Secure=true; SameSite=None; Domain=%s",
                    accessToken, cookieDomain
            );

            String apiKeyCookie = String.format(
                    "apiKey=%s; Path=/; HttpOnly=true; Secure=true; SameSite=None; Domain=%s",
                    apiKey, cookieDomain
            );

            String roleCookie = String.format(
                    "role=%s; Path=/; HttpOnly=true; Secure=true; SameSite=None; Domain=%s",
                    role, cookieDomain
            );

            response.addHeader("Set-Cookie", accessTokenCookie);
            response.addHeader("Set-Cookie", apiKeyCookie);
            response.addHeader("Set-Cookie", roleCookie);
        }

        System.out.println("Set-Cookie 헤더 추가 완료");

        // 실제 응답 헤더 확인
        System.out.println("=== 응답 헤더 확인 ===");
        response.getHeaderNames().forEach(headerName -> {
            if (headerName.equalsIgnoreCase("Set-Cookie")) {
                response.getHeaders(headerName).forEach(headerValue -> {
                    System.out.println("Set-Cookie: " + headerValue);
                });
            }
        });

        // 환경별 리다이렉트 URL 설정
        String redirectUrl;

        if (isLocal) {
            // 로컬 환경: nginx를 통한 리다이렉트
            redirectUrl = String.format(
                    "%s/api/auth/google/callback?token=%s&apiKey=%s&role=%s",
                    frontendUrl, accessToken, apiKey, role
            );
            System.out.println("로컬 환경 - nginx 8888 포트 연결");
        } else {
            // 프로덕션 환경: API 라우트를 통한 쿠키 재설정
            redirectUrl = String.format(
                    "%s/api/auth/google/callback?token=%s&apiKey=%s&role=%s",
                    frontendUrl, accessToken, apiKey, role
            );
            System.out.println("프로덕션 환경 - API 라우트 활용");
        }

        System.out.println("리다이렉트 URL: " + redirectUrl);

        try {
            response.sendRedirect(redirectUrl);
            System.out.println("리다이렉트 완료");
        } catch (Exception e) {
            System.out.println("리다이렉트 실패: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: backend/src/main/kotlin/com/back/back9/global/security/OAuth2SuccessHandler.kt
package com.back.back9.global.security

import com.back.back9.domain.user.repository.UserRepository
import com.back.back9.domain.user.service.UserService
import com.back.back9.domain.wallet.service.WalletService
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Lazy
import org.springframework.security.core.Authentication
import org.springframework.security.web.authentication.AuthenticationSuccessHandler
import org.springframework.stereotype.Component

@Component
class OAuth2SuccessHandler(
    @Lazy private val userService: UserService,
    @Lazy private val walletService: WalletService,
    @Lazy private val userRepository: UserRepository,
    @Value("\${spring.profiles.active:dev}") private val activeProfile: String,
    @Value("\${app.oauth2.frontend-url}") private val frontendUrl: String,
    @Value("\${app.oauth2.cookie-domain:}") private val cookieDomain: String
) : AuthenticationSuccessHandler {

    override fun onAuthenticationSuccess(
        request: HttpServletRequest,
        response: HttpServletResponse,
        authentication: Authentication
    ) {
        println("=== OAuth2SuccessHandler 시작 ===")
        println("활성 프로파일: $activeProfile")

        val securityUser = authentication.principal as SecurityUser
        var user = securityUser.user

        println("인증된 사용자: ${user.userLoginId}")

        if (user.id == null) {
            println("사용자 ID가 null이므로 저장 필요")
            user = userRepository.save(user)
            println("사용자 저장 완료, ID: ${user.id}")
        }

        val accessToken = userService.genAccessToken(user)
        val role = user.role.name
        val apiKey = user.apiKey

        println("액세스 토큰 생성 완료")
        println("사용자 역할: $role")

        if (!walletService.existsByUserId(user.id)) {
            println("지갑이 존재하지 않아 생성")
            walletService.createWallet(user.id)
        } else {
            println("지갑이 이미 존재함")
        }

        println("=== 쿠키 설정 시작 ===")
        println("accessToken: $accessToken")
        println("apiKey: $apiKey")
        println("role: $role")

        val isLocal = activeProfile == "dev"

        fun setCookie(name: String, value: String, extra: String) {
            response.addHeader("Set-Cookie", "$name=$value; Path=/; $extra")
        }

        if (isLocal) {
            println("로컬 환경 - 일반 쿠키 설정")
            val extra = "HttpOnly=true; SameSite=Lax; Max-Age=3600"
            setCookie("accessToken", accessToken, extra)
            setCookie("apiKey", apiKey ?: "", extra)
            setCookie("role", role, extra)
        } else {
            println("프로덕션 환경 - 크로스 도메인 쿠키 설정")
            val extra = "HttpOnly=true; Secure=true; SameSite=None; Domain=$cookieDomain"
            setCookie("accessToken", accessToken, extra)
            setCookie("apiKey", apiKey ?: "", extra)
            setCookie("role", role, extra)
        }

        println("Set-Cookie 헤더 추가 완료")
        println("=== 응답 헤더 확인 ===")
        response.headerNames.filter { it.equals("Set-Cookie", true) }
            .forEach { headerName ->
                response.getHeaders(headerName).forEach { println("Set-Cookie: $it") }
            }

        val redirectUrl = "$frontendUrl/api/auth/google/callback?token=$accessToken&apiKey=$apiKey&role=$role"

        if (isLocal) {
            println("로컬 환경 - nginx 8888 포트 연결")
        } else {
            println("프로덕션 환경 - API 라우트 활용")
        }

        println("리다이렉트 URL: $redirectUrl")

        try {
            response.sendRedirect(redirectUrl)
            println("리다이렉트 완료")
        } catch (e: Exception) {
            println("리다이렉트 실패: ${e.message}")
            e.printStackTrace()
        }
    }
}